x-pack/plugins/security_solution/docs/prebuilt_rules_customization_rfc.md

```diff
@@ -0,0 +1,2397 @@
+
+  - Check if the `prebuilt` field exists in the rule's parameters, and set `immutable` to `true` if it does.
+    - Use case 1: rules that have already been migrated-on-write
+  - If it does not exist, set `immutable` to the current value of the params' `immutable` field.
+    - Use case 2: rules that have not yet been migrated on write
+
+- For `prebuilt`:
+  - If the `prebuilt` field already exists in the internal rule object params, use that value:
+    - Use case 1: prebuilt rules that have already been migrated-on-write
+  - If it does not exist, check the value `immutable` from the rules param:
+    - if it is `true`, create a new `prebuilt` object. This object will only have the `isCustomized` field, set to `false`. The `elasticUpdateDate` date will not be known and must be left undefined.
+      - Use case 2: prebuilt rules that have not been yet migrated-on-write
+    - otherwise, the field should be `undefined`.
+      - Use case 3L: custom rules
+
+##### Rule Management endpoints that will perform normalization-on-read
+
+- All endpoints that also perform migration-on-write, as listed above. All of them use the `internalRuleToAPIResponse` methods to transform the rule before returning the response, so the normalization on read will take place there. For these specific cases, however, the normalization-on-read will take place, but the data stored in ES for each rule should match the response given by the endpoint, i.e. the normalized in-memory value and the stored value of the rule should be the same.
+- **Find Rules - `GET /rules/_find`**: transforms rules in response to `RuleResponse` type with the `internalRuleToAPIResponse` method.
+- **Read Rule - GET /rules:**: transforms rule in response to `RuleResponse` type with the `internalRuleToAPIResponse` method.
+- **Delete Rules - DELETE /rules**: transforms rule in response to `RuleResponse` type with the `internalRuleToAPIResponse` method.
+- **Bulk Actions** - `POST /rules/_bulk_action`: all other action types not mentioned in the previous section will perform normalization-on-read before responding:
+  - Enable
+  - Disable
+  - Delete
+  - Export: see Importing and Exporting Rules section below
+- **Export Rules** - `POST /rules/_export`:
+  - this endpoint is unused via the UI, but might still be used via the public API
+  - same logic as for `_bulk_action` with the `export` action explained in section [Exporting rules](#exporting-rules) below: only normalization-on-read will be performed before writing to the output `ndjson` file.
+
+### Technical implementation of migration-on-write
+
+The logic for the migration of the rule saved objects, which means the determination of the `immutable` and `prebuilt` fields before writing to ES, will be encapsulated in three helper methods:
+
+- `migratePrebuiltSchemaOnRuleCreation` for rule creation
+- `migratePrebuiltSchemaOnRuleUpdate` for rule updates (including patching)
+- `migratePrebuiltSchemaOnRuleBulkEdit` for bulk rule updates
```

Comment:
Agree that the level of detail is too high. Focusing on **what** an API accomplishes from a user's perspective rather than **how** it does it would be helpful for the RFC content.